pkg/webhook/webhook.go

```diff
@@ -199,7 +195,90 @@ func (wh *webhook) validate(ar *v1beta1.AdmissionReview) *v1beta1.AdmissionRespo
 			Status: metav1.StatusFailure, Code: http.StatusForbidden, Reason: "PVC with cloned volumes can't be deleted",
 			Message: fmt.Sprintf("pvc %q has '%v' cloned volume(s)", pvc.Name, len(cStorVolumes.Items)),
 		}
+		return response
+	}
+	return response
+}
+
+// validatePVCCreaterequest validates persistentvolumeclaim(PVC) create request
+func (wh *webhook) validatePVCCreaterequest(req *v1beta1.AdmissionRequest) *v1beta1.AdmissionResponse {
+	response := &v1beta1.AdmissionResponse{}
+	response.Allowed = true
+
+	var pvc corev1.PersistentVolumeClaim
+	err := json.Unmarshal(req.Object.Raw, &pvc)
+	if err != nil {
+		glog.Errorf("Could not unmarshal raw object: %v, %v", err, req.Object.Raw)
+		response.Allowed = false
+		response.Result = &metav1.Status{
+			Status:  metav1.StatusFailure,
+			Code:    http.StatusBadRequest,
+			Reason:  metav1.StatusReasonBadRequest,
+			Message: err.Error(),
+		}
+		return response
+	}
+
+	// If snapshot.alpha.kubernetes.io/snapshot annotation represents the clone pvc
+	// create request
+	snapname := pvc.Annotations["snapshot.alpha.kubernetes.io/snapshot"]
+	if len(snapname) == 0 {
+		return response
+	}
+
+	glog.Infof("AdmissionReview for creating a clone volume Kind=%v, Namespace=%v Name=%v UID=%v patchOperation=%v UserInfo=%v",
+		req.Kind, req.Namespace, req.Name, req.UID, req.Operation, req.UserInfo)
+	// get the snapshot object to get snapshotdata object
+	snapObj, err := wh.snapClientSet.VolumesnapshotV1().VolumeSnapshots(pvc.Namespace).Get(snapname, metav1.GetOptions{})
+	if err != nil {
+		response.Allowed = false
+		response.Result = &metav1.Status{
+			Message: fmt.Sprintf("error retrieving snapshot: %v", err.Error()),
+		}
+		return response
+	}
+	snapDataName := snapObj.Spec.SnapshotDataName
+	glog.V(1).Infof("snapshotdata name: '%s'", snapDataName)
 
+	// get the snapDataObj to get the snapshotdataname
+	snapDataObj, err := wh.snapClientSet.VolumesnapshotV1().VolumeSnapshotDatas().Get(snapDataName, metav1.GetOptions{})
+	if err != nil {
+		response.Allowed = false
+		response.Result = &metav1.Status{
+			Message: fmt.Sprintf("error retrieving snapshot data: %v", err.Error()),
```

Comment:
better to print snapshotdata name, namespace, snapshot name and other details also